### ISSUE TITLE
Using a custom libc derivation

### ISSUE THREAD

**Issue by theoparis** [NixOS/nixpkgs#307600]
## Issue description

I have asked in several places but for some reason no one seems to have a  proper solution to my problem. I want to use a custom libc with nixpkgs (for example, LLVM Libc, bionic or apple's libc built from upstream sources instead of prebuilt nixpkgs packages) and I cannot seem to find a way to override the libc easily by providing a derivation. All I could find is how to override it via a string such as `libc = "musl"` which doesn't help me here.

If this is not supported already, adding support for this could also make #296577 easier. Additionally, this could make porting libc++ or even nix to a new OS simpler.

**Comment by RossComputerGuy** [comment]
Yeah, if being able to specify a custom libc was as simple as passing over a derivation then that could make the process just so much simpler.

**Comment by theoparis** [comment]
This might also allow cosmopolitan libc to be used with nixpkgs and/or LLVM in the future (once https://github.com/jart/cosmopolitan/issues/180 is solved).

**Comment by LuNeder** [comment]
I’ve been wanting this, this would be awesome!!

(tho unfortunately lots of stuff are failing to build in pkgsMusl, we might need to work on that first)

**Comment by theoparis** [comment]
It seems like there *is* [a way to override the libc](https://github.com/MidstallSoftware/nulibc/blob/master/flake.nix#L72). It involves creating an overlay that has a libcCrossChooser. I'm yet to test it with the latest upstream bionic libc though.

**Comment by RossComputerGuy** [comment]
Yes, and #391695 would allow for adding new ABI's via extending `lib`.
